Broker interceptor

mourad opened the pull request:
- Passing in `session` instead of `peer` into broker/dealer calls
- Cleaned up the defaultBroker maps
- introduced BrokerInterceptor

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/spirentorion/turnpike/5)

<!-- Reviewable:end -->

Commit:
Use session instead of peer for broker/dealer calls

Commit:
Introduced 'route' struct instead of multiple maps

Commit:
Introduced BrokerInterceptor

mourad commented:
Leaving this PR opened for now for feedback, however, I realized that the broker_interceptor approach does not scale as the number of organizations increases.  Will need to consider a different approach

<details><summary>

<i>Previously, mourad (Danilo Moura) wrote…</i></summary>



> #### Broker interceptor
> - Passing in `session` instead of `peer` into broker/dealer calls
> - Cleaned up the defaultBroker maps
> - introduced BrokerInterceptor
> 
> </details>

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/spirentorion/turnpike/5#-:-KJHzZBJx6pYSoZeFA4s:-120564551)_

<!-- Sent from Reviewable.io -->

mourad commented:
After thinking about it further,  scaling will be handled by the library user (library has no knowledge of what an organization is -- nor should it).

<details><summary>

<i>Previously, mourad (Danilo Moura) wrote…</i></summary>



> Leaving this PR opened for now for feedback, however, I realized that the broker_interceptor approach does not scale as the number of organizations increases.  Will need to consider a different approach
> </details>

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/spirentorion/turnpike/5#-:-KJJDnTFlqqh01p2V6EP:1946403746)_

<!-- Sent from Reviewable.io -->

djoyner commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_[broker.go, line 20 [r1]](https://reviewable.io:443/reviews/spirentorion/turnpike/5#-KJLK_hqpuCPJXvZxP74:-KJLK_hrLbHjC26HlnLf:1540178298) ([raw file](https://github.com/spirentorion/turnpike/blob/24de77be4705099b9d92d623569b25d018dde41e/broker.go#L20)):_

> ``` Go
> // A super simple broker that matches URIs to Subscribers.
> type defaultBroker struct {
>   routes        map[URI]map[ID]route
> ```

The way this type is defined, as the inner map grows and shrinks the individual `route` elements are going to have to be copied.  Consider `map[URI]map[ID]*route` instead.

---

_[broker_interceptor.go, line 4 [r1]](https://reviewable.io:443/reviews/spirentorion/turnpike/5#-KJLLJFeUnWLcZ0LV2ee:-KJLLJFeUnWLcZ0LV2ef:1912497560) ([raw file](https://github.com/spirentorion/turnpike/blob/24de77be4705099b9d92d623569b25d018dde41e/broker_interceptor.go#L4)):_

> ``` Go
> 
> // BrokerInterceptor is the interface that intercepts broker events
> type BrokerInterceptor interface {
> ```

No issues with the actual implementation but the name is slightly confusing, only because of the clash with the `Interceptor` interface that already exists.  In that interface messages are truly being intercepted and possibly modified.  It doesn't seem like that's what's happening here.  Some alternatives that I can think of:

```
BrokerDistributor
BrokerPublisher
```

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/spirentorion/turnpike/5)_

<!-- Sent from Reviewable.io -->

Commit:
Code Review:
- use pointer to prevent copies when the slice is resized
- rename BrokerInterceptor -> BrokerDistributor

mourad commented:
bump

<details><summary>

<i>Previously, mourad (Danilo Moura) wrote…</i></summary>



> After thinking about it further,  scaling will be handled by the library user (library has no knowledge of what an organization is -- nor should it).
> </details>

---

Review status: 6 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

_[broker.go, line 20 [r1]](https://reviewable.io:443/reviews/spirentorion/turnpike/5#-KJLK_hqpuCPJXvZxP74:-KJMlf5L4uStLDWNm-qJ:-1918224535) ([raw file](https://github.com/spirentorion/turnpike/blob/24de77be4705099b9d92d623569b25d018dde41e/broker.go#L20)):_
<details><summary>

<i>Previously, djoyner (David Joyner) wrote…</i></summary>



> The way this type is defined, as the inner map grows and shrinks the individual `route` elements are going to have to be copied.  Consider `map[URI]map[ID]*route` instead.
> </details>
> hmm .. I should've caught this, thx.

---

_[broker_distributor.go, line 4 [r1]](https://reviewable.io:443/reviews/spirentorion/turnpike/5#-KJLLJFeUnWLcZ0LV2ee:-KJMljF-ZS8gKJIB1per:1854379916) ([raw file](https://github.com/spirentorion/turnpike/blob/24de77be4705099b9d92d623569b25d018dde41e/broker_distributor.go#L4)):_
<details><summary>

<i>Previously, djoyner (David Joyner) wrote…</i></summary>



> No issues with the actual implementation but the name is slightly confusing, only because of the clash with the `Interceptor` interface that already exists.  In that interface messages are truly being intercepted and possibly modified.  It doesn't seem like that's what's happening here.  Some alternatives that I can think of:
> 
> ```
> BrokerDistributor
> BrokerPublisher
> ```
> 
> </details>
> Went with `BrokerDistributor` as I might use it for things other than Publish in the future.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/spirentorion/turnpike/5#-:-KJMltMdaMXgJ4Ahzsb-:677534651)_

<!-- Sent from Reviewable.io -->

djoyner commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/spirentorion/turnpike/5)_

<!-- Sent from Reviewable.io -->
